cpu/stm32l1/stmclk.c

```diff
@@ -0,0 +1,327 @@
+        #define CLOCK_CR_SOURCE_RDY        RCC_CR_HSERDY
+        #define CLOCK_PLL_SOURCE           RCC_CFGR_PLLSRC_HSE
+        #define CLOCK_DISABLE_OTHERS       (RCC_CR_HSION | RCC_CR_MSION)
+
+        #if (CLOCK_USE_PLL == 0)
+            #define CLOCK_CFGR_SW              RCC_CFGR_SW_HSE
+            #define CLOCK_CFGR_SW_RDY          RCC_CFGR_SWS_HSE
+        #endif
+    #endif
+#endif
+
+#if defined(CLOCK_MSI)
+    #define CLOCK_CR_SOURCE            RCC_CR_MSION
+    #define CLOCK_CR_SOURCE_RDY        RCC_CR_MSIRDY
+    #define CLOCK_CFGR_SW              RCC_CFGR_SW_MSI
+    #define CLOCK_CFGR_SW_RDY          RCC_CFGR_SWS_MSI
+    #define CLOCK_DISABLE_OTHERS       (RCC_CR_HSEON | RCC_CR_HSION)
+    #define CLOCK_MSIRANGE             RCC_ICSCR_MSIRANGE_6
+
+    #if (CLOCK_USE_PLL == 1)
+        #error "PLL can't be used with MSI"
+    #endif
+#endif
+
+#if (CLOCK_USE_PLL == 1)
+    #define CLOCK_CFGR_SW              RCC_CFGR_SW_PLL
+    #define CLOCK_CFGR_SW_RDY          RCC_CFGR_SWS_PLL
+#endif
+
+#if (CLOCK_CORECLOCK > 16000000U)
+    #define CORE_VOLTAGE PWR_CR_VOS_0
+#elif (CLOCK_CORECLOCK > 8000000U)
+    #define CORE_VOLTAGE PWR_CR_VOS_1
+#else
+    #define CORE_VOLTAGE (PWR_CR_VOS_1 | PWR_CR_VOS_0)
+#endif
+
+static uint32_t tmpreg;
+static volatile uint32_t clock_source_rdy = 0;
+volatile uint32_t cpu_clock_global;
+volatile uint32_t cpu_ports_number;
+char cpu_clock_source[10] = { 0 };
+
+/**
+ * @brief Configure the clock system of the stm32l1
+ */
+void stmclk_init_sysclk(void)
+{
+    /* Reset the RCC clock configuration to the default reset state(for debug purpose) */
+    /* Set MSION bit */
+    RCC->CR |= RCC_CR_MSION;
+    /* Reset SW, HPRE, PPRE1, PPRE2, MCOSEL and MCOPRE bits */
+    RCC->CFGR &= ~(RCC_CFGR_HPRE | RCC_CFGR_PPRE1 | RCC_CFGR_PPRE2);
+    RCC->CFGR &= ~(RCC_CFGR_MCOSEL | RCC_CFGR_MCOPRE);
+    RCC->CFGR &= ~(RCC_CFGR_SW);
+    /* Reset PLL bits */
+    RCC->CFGR &= ~(RCC_CFGR_PLLSRC | RCC_CFGR_PLLDIV | RCC_CFGR_PLLMUL);
+    /* Reset HSION, HSEON, CSSON and PLLON bits */
+    RCC->CR &= ~(RCC_CR_HSION | RCC_CR_HSEON | RCC_CR_HSEBYP | RCC_CR_CSSON | RCC_CR_PLLON);
+    /* Clear all interrupts */
+    RCC->CIR = 0x0;
+
+    /* SYSCLK, HCLK, PCLK2 and PCLK1 configuration */
+#if defined(CLOCK_HS_MULTI)
+    RCC->CR |= (RCC_CR_HSION | RCC_CR_HSEON);
+
+    /* MCU after reboot or poweron */
+    if (clock_source_rdy == 0) {
+        clock_source_rdy = RCC_CR_HSERDY;
+    }
+
+    volatile int timeout = 0;
+    while (!(RCC->CR & clock_source_rdy)) {
+        timeout++;
+        if (timeout > 10000) {
+            clock_source_rdy = RCC_CR_HSIRDY;
+            timeout = 0;
+        }
+    }
+#else
+    /* Enable high speed clock source */
+    RCC->CR |= CLOCK_CR_SOURCE;
+    /* Wait till the clock source is ready
+     * NOTE: the MCU will stay here forever if you use an external clock source and it's not connected */
+    while (!(RCC->CR & CLOCK_CR_SOURCE_RDY)) {}
```

Comment:
I was about to introduce a timeout here to know if the HSE is ready, otherwise we enable the HSI. Would you mind to introduce something like that?

Reply:
@kYc0o isn't what you are refering introduced at `    volatile int timeout = 0;
    while (!(RCC->CR & clock_source_rdy)) {
        timeout++;
        if (timeout > 10000) {
            clock_source_rdy = RCC_CR_HSIRDY;
            timeout = 0;
        }
    }`

Reply:
Oh yeah, didn't see that code in depth. However I should ask, why only when `defined(CLOCK_HS_MULTI)`? IMHO this should be present whenever any _external_ clock source is being used. So, in your case, are you assuming that HSI can't be used standalone? Otherwise, I think it's worth to show explicitly in the code when only HSI is being used.

There's also a consensus, especially with nucleo boards on which some revisions don't have any HSE or LSE, that we try systematically to use HSE, if it's not present, we switch to internal.

Reply:
Well because when `defined(CLOCK_HS_MULTI)` is not defined CLOCK_CR_SOURCE  would be HSI or HSE, and if CLOCK_HSI is not defined and HSE is not responding for some reason than the only option would be MSI which as far as I know isn't supported as base clock in RIOT, I'm not sure if it could interfere with other drivers (ADC at least wouldn't work); and PLL can't be configured from MSI.

Reply:
What you mean is that only HSI == MSI? As far as I can see in the code for `CLOCK_HS_MULTI` you need both HSI and HSE, and above you're forcing `CLOCK_HS_MULTI` when HSI and HSE are defined. Thus, I need some clarification on how this works and which are the possibilities with boards without HSE and LSE.

Reply:
If CLOCK_HS_MULTI (both HSE and HSI are defined) ir would try to set-up with HSE clock, if timeout it will set-up with HSI.

If CLOCK_HS_MULTI isn't defined, than either CLOCK_HSE or CLOCK_HSI aren't defined for that board. First case scenario you use HSI directly. Case 2 if only HSE is defined for that board then HSI can't be used as a fallback since it isn't defined. 

Only available clock left would be MSI to back-up HSE (since HSI wasn't defined for that board), but since it doesn't work with PLL or ADC it could cause other problems. Therefore blocking doesn't seem that bad. 

When I say a clock is or not defined i mean in periph_cpu or equivalent, not  if it physically exists on the board. Without HSE o LSE it will just go to HSI or MSI if any of those are configured in periph_cpu.

Anyway I guess it's hectic and a little messy. Will give it a second though.

Reply:
Ok I think I got the big picture. Thus, I'd say anyways you always have HSI, so configure it as fallback when HSE is not enabled. Any reason to keep a more complex (yet flexible) config?

Reply:
@kYc0o I'm gonna assume the following things then:
* If HSE try to configure it, if timeout go to HSI
* If HSI use it directly, no timeout needed
* MSI not yet an option as a run clock since there aren't function to properly handle the lower core clock speed this would have.

Reply:
> * If HSE try to configure it, if timeout go to HSI

Discussing offline with @haukepetersen convinced me that this might not be necessary, since it's also *somehow* delaying the startup, so you can just assume the configuration is given by the user who knows if HSE is available.

> * If HSI use it directly, no timeout needed

Agree.
> * MSI not yet an option as a run clock since there aren't function to properly handle the lower core clock speed this would have.

Looking at the implementatin for L4, it seems it's also fallbacking to MSI under certain conditions, though L4 can actually go to 48MHz... Can you just take a look and if needed try to adapt it to this PR?

Reply:
> convinced me that this might not be necessary, since it's also somehow delaying the startup

Why would this delay start-up?

> Looking at the implementation for L4, it seems it's also fall-backing to MSI under certain conditions, though L4 can actually go to 48MHz... Can you just take a look and if needed try to adapt it to this PR?

There are two problems and difference I see here related to L4. 

* L1 MSI goes up to 4 MHz on L1, and usually targeted CORE_CLOCK on nucleo boards is 32MHz,  and CORE_CLOCK is used by other driver to configure their clock speed, which wouldn't be accurate in this case.
*L4 MSI can be used to run the PLL and also ADC threw the SYSCLK, which is not the case for L1.

IMHO, for this to work there should first be support for changing CORE_CLOCK speed for this to work properly since drivers (spi, uart) should be able to adjust to this. And also somehow adc should be disabled when running on MSI in L1. And I think these would be bigger changes since the current clock speed would have to be exposed, and that would start a hole different discussion.

Reply:
> adc should be disabled when running on MSI in L1

Why? ADC on L1 needs HSI to be enabled, but it doesn't care about system clock. On L0 and L4, both HSI and APB clock can be selected as ADC clock source.

Reply:
@olegart you are right, I understood that part wrong.

Reply:
@olegart do you see a problem with leaving the use of MSI as a sysclk source and all it's config for a later PR (this would come with functions to recalculate uart baudrate, xtimer and others) ??